### PR TITLE
machines: Alter memory available helpblock

### PR DESCRIPTION
**Pull request by skobyda**
Won't allow user to input 0 as memory value and use round calculated
memory available downwards.

There are 2 reason why round memory available downwards:
- Spinner in input field rounds maxValue downwards, so this will
bring consistenci into value inputted by spinner vs value inputted
by text input field.
- If we use round, value might be rounded upwards, thus value exceeding
actual memory available might be shown.

### CODE REVIEW

**Comment by skobyda** [comment]
This PR has same commit as #11985, so whichever PR is faster, other one has to drop commit.

**Comment by skobyda** [comment]
Here is preview of an error which shows  after you input 0 memory value but and try to create VM.
The thing is the other helpblock, which says how much memory is available also turns red as error message. To me it seems it's patternfly's FormGroup's thing and I don't know how to change it. @garrett could you help with this problem, or perhaps review design? :)
Screenshot from 2019-06-13 14-53-33

File: test/verify/machineslib.py [hunk 1]
```diff
@@ -1859,7 +1859,7 @@ def tryCreate(self, dialog):
                     b.wait_in_text("#vm-{0}-disks-vda-device".format(name), "disk")
                 else:
                     b.wait_in_text("#vm-{0}-disks-hda-device".format(name), "disk")
-            elif (dialog.storage_pool == 'No Storage' and dialog.sourceType == 'file') or (dialog.sourceType == 'url' and dialog.start_vm):
+            elif dialog.start_vm and (((dialog.storage_pool == 'No Storage' or dialog.storage_size == 0) and dialog.sourceType == 'file') or dialog.sourceType == 'url'):
```
Comment (martinpitt) [inline test/verify/machineslib.py]: This logically seems to be a fixup of the previous commit. If it is, we can squash it on merging, and add the "Closes #"  to the commit message then.

If it's an independent change, please give some explanation in the commit message, and add a "Closes #12063" to the commit message. PRs with more than one commit need to have these so that we can land them with "Rebase & Merge".

File: test/verify/machineslib.py [hunk 2]
```diff
@@ -1530,7 +1533,7 @@ class VmDialog:
 
         def __init__(self, test_obj, name=None,
                      sourceType='file', sourceTypeSecondChoice=None, location='',
-                     memory_size=1, memory_size_unit='GiB',
+                     memory_size=256, memory_size_unit='MiB',
```
Comment (martinpitt) [inline test/verify/machineslib.py]: This changes the memory for all tests. While this is a sensible change to reduce the necessary test resources, it should be an independent commit.

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 3]
```diff
@@ -113,6 +113,10 @@ function validateParams(vmParams) {
         validationFailed['source'] = _("Installation Source should not be empty");
     }
 
+    if (vmParams.memorySize === 0) {
+        validationFailed['memory'] = _("Memory should be more than 0");
```
Comment (martinpitt) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: "more than 0" sounds a bit strange IMHO, and "should" is too weak -- a VM can't function without memory. Perhaps "Memory must not be 0"? @andreasn, any opinion?

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 4]
```diff
@@ -297,24 +301,30 @@ const OSRow = ({ vendor, osInfoList, os, vendors, onValueChanged, validationFail
...
+const MemoryRow = ({ memorySize, memorySizeUnit, nodeMaxMemory, onValueChanged, validationFailed }) => {
+    const validationStateMemory = validationFailed.memory ? 'error' : undefined;
+
     return (
         <React.Fragment>
             <label htmlFor='memory-size' className='control-label'>
                 {_("Memory")}
             </label>
-            <FormGroup bsClass='ct-validation-wrapper' controlId='memory'>
+            <FormGroup validationState={validationStateMemory} bsClass='form-group ct-validation-wrapper' controlId='memory'>
```
Comment (KKoukiou) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: You don't need to add form-group class, it should be added by the FormGroup patternfly component, no?

File: test/verify/machineslib.py [hunk 5]
```diff
@@ -1842,9 +1845,12 @@ def tryCreate(self, dialog):
                 b.wait_present("#vm-{0}-memory".format(name))
                 b.wait_present("#vm-{0}-vcpus".format(name))
 
-            # check memory
-            b.click("#vm-{0}-usage".format(name))
-            b.wait_in_text("tbody.open .listing-ct-body td:nth-child(1) .usage-donut-caption", dialog.getMemoryText())
+            # if OS was specified and memory is 0, virt-install replaces memory value with a value from osinfo
+            # see: virt-manager repo - 5eb47bc00fd75579f94256c272280b26ba0e43f5
+            if dialog.os_vendor != TestMachines.TestCreateConfig.UNSPECIFIED_VENDOR and dialog.memory_size != 0:
```
Comment (KKoukiou) [inline test/verify/machineslib.py]: Since now you do check for memory_size == 0 the create button in this case should be disabled, and the user should not be allowed to create Vms with memory 0.
In this case we don't need to have this extra check here, and this comment either.

File: test/verify/machineslib.py (hunk 1 above)
Comment (skobyda) [reply test/verify/machineslib.py]: I kinda reworked the condition so it's more accurate. Not exactly a fixup, but I can tell where I should have done things differently:
- 9b5ed7da4cfbab059a94de72c62576088b09b021 (I should have added start_vm for whole condition)
- a063c8bf0f5a36ba7720554e6be57cbb711544df (I should have added storage_size == 0)
both commits are in different PRs.

Should I therefore split this commit into two commits, and each will close it's relative PR where change should have been made?

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx (hunk 4 above)
Comment (skobyda) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: It's weird, but it seems once you define your own bsClass, you lost form-group added by Patternfly.
e.g. Name Row doesn't specify it's own bsClass, and it does have form-group added. Storage Size row specified it's own bsClass, and it looses form-group.

Comment (KKoukiou) [reply pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: Right, noticed just. I just tried to add it and I don't see any visual change but since we have it everywhere let's keep it here.

File: pkg/machines/components/create-vm-dialog/createVmDialog.jsx [hunk 6]
```diff
@@ -152,6 +152,10 @@ function validateParams(vmParams) {
         validationFailed['source'] = _("Installation Source should not be empty");
     }
 
+    if (vmParams.memorySize === 0) {
+        validationFailed['memory'] = _("Memory must be more than 0");
```
Comment (KKoukiou) [inline pkg/machines/components/create-vm-dialog/createVmDialog.jsx]: Ups this is a reminder issue, @martinpitt asked you to change this and the conversation was marked as resolved.
s/must\ be\ more\ than/must not be/g

**Comment by KKoukiou** [comment]
@skobyda there are a bunch of same failures on testCreate for virsh provider, can you take a look?
For example:
https://209.132.184.41:8493/logs/pull-12063-20190625-131313-dbdab5ee-verify-ubuntu-1804/TestMachinesVirsh-testCreate-ubuntu-1804-127.0.0.2-2501-FAIL.png

**Comment by skobyda** [comment]
So I was wrong and it seems that on some distros (debian and ubuntu) you cannot **run** a VM with memory value same as total memory on host. So I deleted autostart option which I previously added.

**Comment by skobyda** [comment]
To give more info, if the total memory value (shown as available in our helpblock) is given as memory value to VM and VM is started, not all of that memory is propably allocated to VM, and rest is supplemented by swap memory. On rhel and fedora bots, there was enough swap memory for tests to pass, on debian and ubuntu there was not enough swap.

This however makes thing if helpblock text (Up to xxx memory available on the host), since not all of that memory is actually available, and if you try to create VM with that memory, it might work only because there is enought of free memory on swap, but sometimes might not work. So perhaps let's change that that helpblock text, to something like (xxx of total memory on host)? To make it more verbose that not all of that memory is actually there available for user and VM to use. But we cannot give precise number. @andreasn what do you think?

**Comment by KKoukiou** [comment]
> To give more info, if the total memory value (shown as available in our helpblock) is given as memory value to VM and VM is started, not all of that memory is propably allocated to VM, and rest is supplemented by swap memory. On rhel and fedora bots, there was enough swap memory for tests to pass, on debian and ubuntu there was not enough swap.
> 
> This however makes thing if helpblock text (Up to xxx memory available on the host), since not all of that memory is actually available, and if you try to create VM with that memory, it might work only because there is enought of free memory on swap, but sometimes might not work. So perhaps let's change that that helpblock text, to something like (xxx of total memory on host)? To make it more verbose that not all of that memory is actually there available for user and VM to use. But we cannot give precise number. @andreasn what do you think?

@skobyda Good catch but let's not block this PR on this. All these wordings make clear that host actually has this X amount of memory. People tend to be aware that one should probably not use the full amount of memory of the host for some guest.

**Comment by skobyda** [comment]
None of the test failures are machines related.
@KKoukiou @martinpitt Since your last review, I have only changed single test case parameter to autostart_vm = false, as explained above.

**Comment by andreasn** [comment]
Sorry for being late to the party!
I think the string "up to..." is OK for now, and I hope people will figure it out.
But we can fix it in the future if it turns out it's a problem.
